### PR TITLE
Error fixes

### PR DESCRIPTION
**Pull request by saldanaj97**
# Enhanced Error Handling and Resilience for LeetCode Submission Fetching

## Problem
The current implementation crashes the entire application when encountering any submission fetch failure. This is particularly problematic because:
1. A single failed submission causes the entire sync process to fail
2. The retry mechanism wasn't handling all error cases effectively
3. Error messages weren't providing enough context for debugging

## Solution
Enhanced the error handling and resilience of the LeetCode submission fetching process with the following improvements:

### 1. Graceful Failure Handling
- Changed behavior to skip individual failed submissions instead of crashing the entire process
- Only returns an error if ALL submissions fail to fetch
- Added progress tracking (X/Y submissions fetched successfully)

### 2. Improved Retry Logic
- Added detection and retry for null responses (`body.Data.Details == nil`)
- Added specific handling for different types of failures:
  - Network errors
  - Null responses
  - Empty code responses
- Each type of failure now has its own retry counter and clear logging

### 3. Better Error Context
- Wrapped errors with additional context using `fmt.Errorf`
- Added more descriptive error messages
- Improved logging for better debugging capabilities

## Testing
The changes have been tested with:
- Normal submission fetching
- Cases where some submissions fail

## Impact
These changes will:
- Reduce application crashes
- Provide better visibility into fetching issues
- Improve the overall reliability of the sync process

### CODE REVIEW

**Comment by ahmed-e-abdulaziz** [comment]
Thanks for your contribution, @saldanaj97. I have some comments that I will add but all in all this looks great.

File: code/leetcode.go [hunk 1]
```diff
@@ -24,7 +24,7 @@ var submissionListQuery string
 var userProgressQuestionListQuery string
 
 const maxRetry = 25                 // LeetCode API can fail A LOT :( It requires a ton of retries when it fails
-const backoffTime = 1 * time.Second // 1 second to avoid keep using LeetCode API when it fails
+const backoffTime = 2 * time.Second // 1 second to avoid keep using LeetCode API when it fails
```
Comment (ahmed-e-abdulaziz) [inline code/leetcode.go]: Was 1 second too fast when you tried it?

File: code/leetcode.go [hunk 2]
```diff
@@ -91,48 +116,58 @@ func (lc leetcode) fetchQuestions() ([]lcQuestion, error) {
...
-		return lcSumbissionOverview{}, errors.New("couldn't fetch any submissions for question with title slug: " + titleSlug)
+		return lcSumbissionOverview{}, fmt.Errorf("no submissions found for question: %s", titleSlug)
 	}
 	return body.Data.LCSubmissionList.LCSubmissions[0], nil // we only need the lastest submission
 }
 
 // Fetches submission's code using the leetcode's submission id
 // Uses LC's GraphQl query that's called submissionDetails
-//
-// id is usually a string of numbers fetched earlier by fetchSubmissionOverview
```
Comment (ahmed-e-abdulaziz) [inline code/leetcode.go:111]: Did you remove this because it was too obvious what the ID and the error return were?

File: code/leetcode.go [hunk 3]
```diff
@@ -36,34 +36,59 @@ func NewLeetCode(cfg config.Config, leetcodeGraphqlUrl string) leetcode {
...
-		}
-		submissions[idx] = Submission{question.FrontendId, question.Title, question.TitleSlug, question.LastSubmittedAt, lcSubmission.Lang, code}
-	}
-	log.Print("Fetched submissions successfully\n==============\n")
-	return submissions, nil
+    log.Println("\n==============\nFetching submissions next")
+    questions, err := lc.fetchQuestions()
+    if err != nil {
+        log.Printf("Error fetching questions: %v\n", err)
+        return nil, fmt.Errorf("question fetching error: %w", err)
```
Comment (ahmed-e-abdulaziz) [inline code/leetcode.go]: No need to log `err` and then also return it. Instead you can log the error and only return a generic error message to avoid logging the same error twice (here and in the handler)

**Review by ahmed-e-abdulaziz — DISMISSED** [review]
Great PR, approved with a small nitpick. It just needs some tests too

File: code/leetcode.go [hunk 4]
```diff
@@ -91,48 +116,58 @@ func (lc leetcode) fetchQuestions() ([]lcQuestion, error) {
...
+        if retry < maxRetry {
+            log.Printf("Null response, retry %d/%d after %v\n", retry+1, maxRetry, backoffTime)
+            time.Sleep(backoffTime)
+            return lc.fetchSubmissionCode(id, retry+1)
+        }
+        return "", fmt.Errorf("max retries reached, consistently getting null response for submission %s", id)
+    }
+
+    if len(body.Data.Details.Code) == 0 {
+        if retry < maxRetry {
```
Comment (ahmed-e-abdulaziz) [inline code/leetcode.go:166]: I never saw this happen, as in having details but with no code. But we don't have any documentation for LeetCode API so it's better to be safe than sorry

**Review by ahmed-e-abdulaziz — CHANGES_REQUESTED** [review]
Great PR!. It just needs some tests for the new negative scenarios. Please add them and check my other comments

**Comment by ahmed-e-abdulaziz** [comment]
BTW ignore the failing github workflow, I am yet to add a separate flow for PRs

File: code/leetcode.go (hunk 1 above)
Comment (saldanaj97) [reply code/leetcode.go]: No, 1 second works fine. It was a precautionary measure to slow down the retries just in case a rate limit was being reached. I could not find any info on the Leetcode API rate limit but setting backOffTIme to 1 second still yielded 199/200 solutions uploaded so I will revert backOffTime to 1 second.

File: code/leetcode.go (hunk 2 above)
Comment (saldanaj97) [reply code/leetcode.go:111]: I do believe the id is pretty self explanatory but I have added the error return comment back as that is helpful in explaining what we are expecting to be returned.

File: code/leetcode.go (hunk 4 above)
Comment (saldanaj97) [reply code/leetcode.go:166]: I also did not end up encountering this personally but this was intended as a temporary fix for [Issue #1 ](https://github.com/ahmed-e-abdulaziz/glsync/issues/1) as there is currently no more information about the error from the user apart from a screenshot which shows us that the user successfully pulled the question name and number but not the code/submission. 

I will add tests for this case.

**Review by saldanaj97 — COMMENTED** [review]
Changes from comments have been applied. Tests for the new functionality have been added and passed.

**Review by ahmed-e-abdulaziz — APPROVED** [review]
LGTM 🚀

**Comment by ahmed-e-abdulaziz** [comment]
BTW, another thing I had to fix afterward: @saldanaj97, you were using spaces for indentation.
In Go, the standard is to use tabs instead of spaces https://stackoverflow.com/questions/19094704/indentation-in-go-tabs-or-spaces it's a small thing but thought to mention it as it can be helpful to stick to the standard code convention

**Comment by saldanaj97** [comment]
Thank you for the feedback! I'm still rather new to Go so I was not aware of the standard practice but I greatly appreciate you letting me know for future projects.

**Comment by ahmed-e-abdulaziz** [comment]
No worries. Honestly, you did great in this PR. If you use VSCode, install the Go extension, which should install `gopls`. That will take care of all of Go related things in VS Code.
Thanks again for your contribution 🙏
